doc/config/veilid-server-config.md
---
title: Veilid Server Configuration
keywords:
- config
- veilid-server
status: Draft
---
# Veilid Server Configuration

## Configuration File

`veilid-server` may be run using configuration from both command-line arguments
and the `veilid-server.conf` file.

## Global Directives

| Directive                    | Description                           |
| ---------------------------- | ------------------------------------- |
| [daemon](#daemon)            | Run `veilid-server` in the background |
| [client\_api](#client_api)   |                                       |
| [auto\_attach](#auto_attach) |                                       |
| [logging](#logging)          |                                       |
| [testing](#testing)          |                                       |
| [core](#core)                |                                       |


### daemon

```yaml
daemon:
    enabled: false
```

### client_api

```yaml
client_api:
    enabled: true
    listen_address: 'localhost:5959'
```

| Parameter                                    | Description |
| -------------------------------------------- | ----------- |
| [enabled](#client_apienabled)                |             |
| [listen\_address](#client_apilisten_address) |             |

#### client\_api:enabled

**TODO**

#### client\_api:listen\_address

**TODO**

### auto\_attach

```yaml
auto_attach: true
```

### logging

```yaml
logging:
    system:
        enabled: false
        level: 'info'
    terminal:
        enabled: true
        level: 'info'
    file:
        enabled: false
        path: ''
        append: true
        level: 'info'
    api:
        enabled: true
        level: 'info'
    otlp:
        enabled: false
        level: 'trace'
        grpc_endpoint: 'localhost:4317'
```

| Parameter                    | Description |
| ---------------------------- | ----------- |
| [system](#loggingsystem)     |             |
| [terminal](#loggingterminal) |             |
| [file](#loggingfile)         |             |
| [api](#loggingapi)           |             |
| [otlp](#loggingotlp)         |             |

#### logging:system

```yaml
system:
    enabled: false
    level: 'info'
```

#### logging:terminal

```yaml
terminal:
    enabled: true
    level: 'info'
```

#### logging:file

```yaml
file:
    enabled: false
    path: ''
    append: true
    level: 'info'
```

#### logging:api

```yaml
api:
    enabled: true
    level: 'info'
```

#### logging:otlp

```yaml
otlp:
    enabled: false
    level: 'trace'
    grpc_endpoint: 'localhost:4317'
```

### testing

```yaml
testing:
    subnode_index: 0
    subnode_count: 1
```

### core

| Parameter                                | Description |
| ---------------------------------------- | ----------- |
| [protected\_store](#coreprotected_store) |             |
| [table\_store](#coretable_store)         |             |
| [block\_store](#block_store)             |             |
| [network](#corenetwork)                  |             |

#### core:protected\_store

```yaml
protected_store:
    allow_insecure_fallback: true
    always_use_insecure_storage: true
    directory: '%DIRECTORY%'
    delete: false
```

#### core:table\_store

```yaml
table_store:
    directory: '%TABLE_STORE_DIRECTORY%'
    delete: false
```

#### core:block\_store

```yaml
block_store:
    directory: '%BLOCK_STORE_DIRECTORY%'
    delete: false
```

#### core:network

```yaml
network:
    connection_initial_timeout_ms: 2000
    connection_inactivity_timeout_ms: 60000
    max_connections_per_ip4: 32
    max_connections_per_ip6_prefix: 32
    max_connections_per_ip6_prefix_size: 56
    max_connection_frequency_per_min: 128
    client_allowlist_timeout_ms: 300000
    reverse_connection_receipt_time_ms: 5000
    hole_punch_receipt_time_ms: 5000
    network_key_password: null
    disable_capabilites: []
    node_id: null
    node_id_secret: null
    upnp: true
    detect_address_changes: true
    enable_local_peer_scope: false
    restricted_nat_retries: 0
```

| Parameter                                   | Description |
| ------------------------------------------- | ----------- |
| [routing\_table](#corenetworkrouting_table) |             |
| [rpc](#corenetworkrpc)                      |             |
| [dht](#corenetworkdht)                      |             |
| [tls](#corenetworktls)                      |             |
| [application](#corenetworkapplication)      |             |
| [protocol](#corenetworkprotocol)            |             |

#### core:network:routing\_table

```yaml
routing_table:
    node_id: null
    node_id_secret: null
    bootstrap: ['bootstrap-v1.veilid.net']
    bootstrap_keys: ['VLD0:Vj0lKDdUQXmQ5Ol1SZdlvXkBHUccBcQvGLN9vbLSI7k','VLD0:QeQJorqbXtC7v3OlynCZ_W3m76wGNeB5NTF81ypqHAo','VLD0:QNdcl-0OiFfYVj9331XVR6IqZ49NG-E18d5P7lwi4TA']
    limit_over_attached: 64
    limit_fully_attached: 32
    limit_attached_strong: 16
    limit_attached_good: 8
    limit_attached_weak: 4
```

#### core:network:rpc

```yaml
rpc:
    concurrency: 0
    queue_size: 1024
    max_timestamp_behind_ms: 10000
    max_timestamp_ahead_ms: 10000
    timeout_ms: 5000
    max_route_hop_count: 4
    default_route_hop_count: 1
```

#### core:network:dht

```yaml
dht:
    max_find_node_count: 20
    resolve_node_timeout_ms: 10000
    resolve_node_count: 1
    resolve_node_fanout: 4
    get_value_timeout_ms: 10000
    get_value_count: 3
    get_value_fanout: 4
    set_value_timeout_ms: 10000
    set_value_count: 5
    set_value_fanout: 4
    min_peer_count: 20
    min_peer_refresh_time_ms: 60000
    validate_dial_info_receipt_time_ms: 2000
    local_subkey_cache_size: 128
    local_max_subkey_cache_memory_mb: 256
    remote_subkey_cache_size: 1024
    remote_max_records: 65536
    remote_max_subkey_cache_memory_mb: %REMOTE_MAX_SUBKEY_CACHE_MEMORY_MB%
    remote_max_storage_space_mb: 0
    public_watch_limit: 32
    member_watch_limit: 8
    max_watch_expiration_ms: 600000
```

#### core:network:tls

```yaml
tls:
    certificate_path: '%CERTIFICATE_PATH%'
    private_key_path: '%PRIVATE_KEY_PATH%'
    connection_initial_timeout_ms: 2000
```

#### core:network:application

```yaml
application:
    https:
        enabled: false
        listen_address: ':5150'
        path: 'app'
        # url: 'https://localhost:5150'
    http:
        enabled: false
        listen_address: ':5150'
        path: 'app'
        # url: 'http://localhost:5150'
```

#### core:network:protocol

```yaml
protocol:
    udp:
        enabled: true
        socket_pool_size: 0
        listen_address: ':5150'
        # public_address: ''
    tcp:
        connect: true
        listen: true
        max_connections: 32
        listen_address: ':5150'
        #'public_address: ''
    ws:
        connect: true
        listen: true
        max_connections: 16
        listen_address: ':5150'
        path: 'ws'
        # url: 'ws://localhost:5150/ws'
    wss:
        connect: true
        listen: false
        max_connections: 16
        listen_address: ':5150'
        path: 'ws'
        # url: ''
```


doc/config/veilid-dev-node-config.md
# Veilid Server
# =============
#
# Dev Node Configuration
#
# -----------------------------------------------------------

---
logging:
  system:
    enabled: true
    level: debug
  api:
    enabled: true
    level: debug
  terminal:
    enabled: false
core:
  capabilities:
    disable: ['APPM']
  network:
    upnp: false
    dht:
      min_peer_count: 10
    detect_address_changes: false
    routing_table:
      bootstrap: ['bootstrap.<your.domain>']
    network_key_password: '<your-chosen-passkey>'
  protected_store:
    insecure_fallback_directory: '/var/db/veilid-server/protected_store'
  table_store:
    directory: '/var/db/veilid-server/table_store'
  block_store:
    directory: '/var/db/veilid-server/block_store'

CHANGELOG.md
**UNRELEASED**

- Update capnproto version to 1.1.0
- *BREAKING API CHANGE*:
  - watch_dht_values() now returns a bool rather than an expiration timestamp. Expiration renewal is now managed by veilid-core internally. Apps no longer need to renew watches!
  - inspect_dht_record() and cancel_dht_watch() now take an Option<ValueSubkeyRangeSet> instead of just a ValueSubkeyRangeSet, to make things easier for automatic binding generation someday and to remove ambiguities about the semantics of the default empty set.
  - DHTRecordReport now uses a Vec<Option<ValueSubkey>> for seq lists, rather than using the 'ValueSubkey::MAX' sentinel value (0xFFFFFFFF) to represent a missing subkey
  - Renamed config structs to better describe their purpose, and remove "Inner" from a struct that's being exposed via the API. ([!402](https://gitlab.com/veilid/veilid/-/merge_requests/402))
    - `VeilidConfig` -> `VeilidStartupOptions`
    - `VeilidConfigInner` -> `VeilidConfig`
  - Gate insecure capabilities behind the new `footgun` feature flag ([#394](https://gitlab.com/veilid/veilid/-/issues/394)), which is disabled by default. ([!400](https://gitlab.com/veilid/veilid/-/merge_requests/400))
    - Calling `app_call` or `app_message` with a `NodeId` target will throw an error. Use a `PrivateRoute` target instead.
    - Creating an `Unsafe` routing context will throw and error. Use `Safe` routing context instead.
    - Any AppCall or AppMessage sent from a direct NodeId will not be received.

- veilid-core:
  - **Security** Signed bootstrap v1 added which closes #293: https://gitlab.com/veilid/veilid/-/issues/293
  - Allow shutdown even if tables are closed
  - New, more robust, watchvalue implementation
  - Consensus is now counted from the nodes closest to the key, excluding attempts that have failed, but including new nodes that show up, requiring N out of the M closest nodes to have      succeeded and all have been attempted.
  - Watching a node now also triggers an background inspection+valueget to detect if values have changed online
  - Fanout queue disqualifaction for distance-based rejections reimplemented
  - Local rehydration implemented. DHT record subkey data that does not have sufficient consensus online is re-pushed to keep it alive when records are opened.
  - Direct bootstrap v0 now filters out Relayed nodes correctly
  - Closed issue #400: https://gitlab.com/veilid/veilid/-/issues/400
  - Closed issue #377: https://gitlab.com/veilid/veilid/-/issues/377
  - Add the `veilid_features()` API, which lists the compile-time features that were enabled when `veilid-core` was built (available in language bindings as well). ([!401](https://gitlab.com/veilid/veilid/-/issues/400))
  - When `veilid-core` starts up, log the version number, and the compile-time features that were enabled when it was built. ([!401](https://gitlab.com/veilid/veilid/-/issues/400))
  - Closed issue #448: https://gitlab.com/veilid/veilid/-/issues/448
  - Add background flush for routing table and route spec store, to address issue #449: https://gitlab.com/veilid/veilid/-/issues/449

- veilid-flutter:
  - Bindings updated for API changes
  - Corrosion version in cmake build for linux and windows updated to 0.5.1: https://gitlab.com/veilid/veilid/-/issues/447
  - Expose the isShutdown API: https://gitlab.com/veilid/veilid/-/merge_requests/392

- veilid-python:
  - Fix type assertion bug in watch_dht_values
  - Update watchvalue integration tests
  - Expose the is_shutdown API: https://gitlab.com/veilid/veilid/-/merge_requests/392

- veilid-server:
  - Put tokio-console behind a feature flag. Closed issue #274: https://gitlab.com/veilid/veilid/-/issues/274
  - Fixed 'daemon' mode `-d` option. Closed issue #360: https://gitlab.com/veilid/veilid/-/issues/360

- veilid-wasm:
  - **Breaking** Properly generate TypeScript types for `ValueSubkeyRangeSet`, which would previously resolve to `any`. This is breaking since it can cause type errors to correctly surface in existing applications. ([!397](https://gitlab.com/veilid/veilid/-/merge_requests/397))
  - **Breaking** `startupCore()` and `defaultConfig()` now use config objects instead of stringified JSON.
    - `veilidClient.startupCore(callback, JSON.stringify(config))` now becomes `veilidClient.startupCore(callback, config)`. ([!402](https://gitlab.com/veilid/veilid/-/merge_requests/402))
    - `JSON.parse(veilidClient.defaultConfig())` is now `veilidClient.defaultConfig()`
    - The `VeilidConfigInner` type is now `VeilidConfig`.
  - Expose the isShutdown API: https://gitlab.com/veilid/veilid/-/merge_requests/392

- CI:
  - Ensure Cargo.lock is up-to-date during CI pipelines

**Changed in Veilid 0.4.4**

- veilid-core:
  - Improved termination conditions for DiscoveryContext, faster shutdown acknowledgment
  - Network class discovery from 25s to 5s in worst case
  - Moved some futures to the heap with Box::pin to reduce risk of stack overflows
  - Restore punishment for FailedToVerifySenderPeerInfo
  - Convert inbound relay loop notification from an error to a NetworkResult
  - Relays were being picked before we had good stats on their performance. Relays are now optimized to be within the top 25% of p90 latencies
  - LatencyStats now have median 80% average, p90, p75, and p50 latency numbers
  - RPC Answer stats now are split by 'ordered' and 'unordered' protocols
  - Unordered protocols are given a higher tolerance for lost answers before being marked 'unreliable'
  - Fix race condition in TTL code for udp hole punching
  - Correct use of IPV6_UNICAST_HOPS socket option instead of IP_TTL for ipv6 sockets
  - Move relaying to a relay worker pool
  - Connection table and RPC locking improvements
  - Remove TCP connection retries
  - Deadlock bugfix in relay code
  - 'Finding more nodes' tasks now use PreferOrdered sequencing
  - Fix deadlock in veilid_api duration testing
  - Correct offline_subkey_writes inflight reporting

veilid-flutter/example/lib/config.dart
import 'package:veilid/veilid.dart';
import 'package:flutter/foundation.dart' show kIsWeb;
import 'package:path_provider/path_provider.dart';
import 'package:path/path.dart' as p;

Future<VeilidConfig> getDefaultVeilidConfig() async {
  return VeilidConfig(
    programName: "Veilid Plugin Test",
    namespace: "",
    capabilities: VeilidConfigCapabilities(
      protocolUDP: !kIsWeb,
      protocolConnectTCP: !kIsWeb,
      protocolAcceptTCP: !kIsWeb,
      protocolConnectWS: true,
      protocolAcceptWS: !kIsWeb,
      protocolConnectWSS: true,
      protocolAcceptWSS: false,
    ),
    protectedStore: VeilidConfigProtectedStore(
      allowInsecureFallback: false,
      alwaysUseInsecureStorage: false,
      insecureFallbackDirectory: "",
      delete: false,
    ),
    tableStore: VeilidConfigTableStore(
      directory: kIsWeb
          ? ""
          : p.join((await getApplicationSupportDirectory()).absolute.path,
              "table_store"),
      delete: false,
    ),
    blockStore: VeilidConfigBlockStore(
      directory: kIsWeb
          ? ""
          : p.join((await getApplicationSupportDirectory()).absolute.path,
              "block_store"),
      delete: false,
    ),
    network: VeilidConfigNetwork(
      connectionInitialTimeoutMs: 2000,
      connectionInactivityTimeoutMs: 60000,
      maxConnectionsPerIp4: 32,
      maxConnectionsPerIp6Prefix: 32,
      maxConnectionsPerIp6PrefixSize: 56,
      maxConnectionFrequencyPerMin: 128,
      clientWhitelistTimeoutMs: 300000,
      reverseConnectionReceiptTimeMs: 5000,
      holePunchReceiptTimeMs: 5000,
      nodeId: null,
      nodeIdSecret: null,
      bootstrap: kIsWeb
          ? ["ws://bootstrap.dev.veilid.net:5150/ws"]
          : ["bootstrap.dev.veilid.net"],
      bootstrapNodes: [],
      routingTable: VeilidConfigRoutingTable(
        limitOverAttached: 64,
        limitFullyAttached: 32,
        limitAttachedStrong: 16,
        limitAttachedGood: 8,
        limitAttachedWeak: 4,
      ),
      rpc: VeilidConfigRPC(
        concurrency: 0,
        queueSize: 1024,
        maxTimestampBehindMs: 10000,
        maxTimestampAheadMs: 10000,
        timeoutMs: 10000,
        maxRouteHopCount: 4,
        defaultRouteHopCount: 1,
      ),
      dht: VeilidConfigDHT(
        resolveNodeTimeoutMs: null,
        resolveNodeCount: 20,
        resolveNodeFanout: 3,
        maxFindNodeCount: 20,
        getValueTimeoutMs: null,
        getValueCount: 20,
        getValueFanout: 3,
        setValueTimeoutMs: null,
        setValueCount: 20,
        setValueFanout: 5,
        minPeerCount: 20,
        minPeerRefreshTimeMs: 2000,
        validateDialInfoReceiptTimeMs: 2000,
      ),
      upnp: true,
      detectAddressChanges: true,
      restrictedNatRetries: 0,
      tls: VeilidConfigTLS(
        certificatePath: "",
        privateKeyPath: "",
        connectionInitialTimeoutMs: 2000,
      ),
      application: VeilidConfigApplication(
          https: VeilidConfigHTTPS(
            enabled: false,
            listenAddress: "",
            path: "",
            url: null,
          ),
          http: VeilidConfigHTTP(
            enabled: false,
            listenAddress: "",
            path: "",
            url: null,
          )),
      protocol: VeilidConfigProtocol(
        udp: VeilidConfigUDP(
          enabled: !kIsWeb,
          socketPoolSize: 0,
          listenAddress: "",
          publicAddress: null,
        ),
        tcp: VeilidConfigTCP(
          connect: !kIsWeb,
          listen: !kIsWeb,
          maxConnections: 32,
          listenAddress: "",
          publicAddress: null,
        ),
        ws: VeilidConfigWS(
          connect: true,
          listen: !kIsWeb,
          maxConnections: 16,
          listenAddress: "",
          path: "ws",
          url: null,
        ),
        wss: VeilidConfigWSS(
          connect: true,
          listen: false,
          maxConnections: 16,
          listenAddress: "",
          path: "ws",
          url: null,
        ),
      ),
    ),
  );
}
